File: apps/Pose-Motion/iOS/PoseMotion/Overlay/VideoFitMapper.swift
import CoreGraphics

/// Aspect-fit math shared by every overlay: where the video frame lands inside the view.
enum VideoFitMapper {
    static func fitRect(content: CGSize, in container: CGSize) -> CGRect {
        guard content.width > 0, content.height > 0,
              container.width > 0, container.height > 0 else { return .zero }
        let scale = min(container.width / content.width, container.height / content.height)
        let size = CGSize(width: content.width * scale, height: content.height * scale)
        return CGRect(
            x: (container.width - size.width) / 2,
            y: (container.height - size.height) / 2,
            width: size.width,
            height: size.height
        )
    }

    /// Normalized (0..1) frame point → view point.
    static func point(_ p: CGPoint, in fit: CGRect) -> CGPoint {
        CGPoint(x: fit.minX + p.x * fit.width, y: fit.minY + p.y * fit.height)
    }

    /// Normalized (0..1) frame rect → view rect.
    static func rect(_ r: CGRect, in fit: CGRect) -> CGRect {
        CGRect(
            x: fit.minX + r.minX * fit.width,
            y: fit.minY + r.minY * fit.height,
            width: r.width * fit.width,
            height: r.height * fit.height
        )
    }
}

File: apps/Pose-Motion/Android/app/src/main/java/ai/zetic/demo/posemotion/ui/overlay/VideoFitMapper.kt
package ai.zetic.demo.posemotion.ui.overlay

import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Rect
import androidx.compose.ui.geometry.Size

/** Aspect-fit math shared by every overlay: where the video frame lands in the canvas. */
object VideoFitMapper {
    fun fitRect(contentW: Float, contentH: Float, container: Size): Rect {
        if (contentW <= 0 || contentH <= 0 || container.width <= 0 || container.height <= 0) {
            return Rect.Zero
        }
        val scale = minOf(container.width / contentW, container.height / contentH)
        val w = contentW * scale
        val h = contentH * scale
        val left = (container.width - w) / 2
        val top = (container.height - h) / 2
        return Rect(left, top, left + w, top + h)
    }

    /** Normalized (0..1) frame point → canvas point. */
    fun point(x: Float, y: Float, fit: Rect): Offset =
        Offset(fit.left + x * fit.width, fit.top + y * fit.height)

    /** Normalized (0..1) frame rect → canvas rect. */
    fun rect(r: android.graphics.RectF, fit: Rect): Rect = Rect(
        fit.left + r.left * fit.width,
        fit.top + r.top * fit.height,
        fit.left + r.right * fit.width,
        fit.top + r.bottom * fit.height,
    )
}
